doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,422 @@
+
+### Usage
+
+```rust
+// Open a txns shard, initializing it if necessary.
+let mut txns = TxnsHandle::open(0u64, client, ShardId::new()).await;
+
+// Register data shards to the txn set.
+let (d0, d1) = (ShardId::new(), ShardId::new());
+txns.register(d0, 1u64).await.expect("not previously initialized");
+txns.register(d1, 2u64).await.expect("not previously initialized");
+
+// Commit a txn. This is durable if/when the `commit_at` succeeds, but reads
+// at the commit ts will _block_ until after the txn is applied. Users are
+// free to pass up the commit ack (e.g. to pgwire) to get a bit of latency
+// back. NB: It is expected that the txn committer will run the apply step,
+// but in the event of a crash, neither correctness nor liveness depend on
+// it.
+let mut txn = txns.begin();
+txn.write(&d0, vec![0], 1);
+txn.write(&d1, vec![1], -1);
+txn.commit_at(&mut txns, 3).await.expect("ts 3 available")
+    // And make it available to reads by applying it.
+    .apply(&mut txns).await;
+
+// Commit a contended txn at a higher timestamp. Note that the upper of `d1`
+// is also advanced by this.
+let mut txn = txns.begin();
+txn.write(&d0, vec![2], 1);
+txn.commit_at(&mut txns, 3).await.expect_err("ts 3 not available");
+txn.commit_at(&mut txns, 4).await.expect("ts 4 available")
+    .apply(&mut txns).await;
+
+// Read data shard(s) at some `read_ts`.
+let updates = d1_read.snapshot_and_fetch(
+    vec![txns.read_cache().to_data_inclusive(&d1, 4).unwrap()].into()
+).await.unwrap();
+```
+
+### Writes
+
+The structure of the txns shard is `(ShardId, Vec<u8>)` updates.
+
+The core mechanism is that a txn commits a set of transmittable persist _batch
+handles_ as `(ShardId, <opaque blob>)` pairs at a single timestamp. This
+contractually both commits the txn and advances the logical upper of _every_
+data shard (not just the ones involved in the txn).
+
+Example:
+
+```text
+// A txn to only d0 at ts=1
+(d0, <opaque blob A>, 1, 1)
+// A txn to d0 (two blobs) and d1 (one blob) at ts=4
+(d0, <opaque blob B>, 4, 1)
+(d0, <opaque blob C>, 4, 1)
+(d1, <opaque blob D>, 4, 1)
+```
+
+However, the new commit is not yet readable until the txn apply has run, which
+is expected to be promptly done by the committer, except in the event of a
+crash. This, in ts order, moves the batch handles into the data shards with a
+`compare_and_append_batch` (similar to how the multi-worker persist_sink works).
+
+Once apply is run, this is noted in the txns shard by retracting the update
+adding the batch. As a result, the contents of the txns shard at any given
+timestamp is exactly the set of outstanding apply work.
+
+Example (building on the above):
+
+```text
+// Apply for the first txn at ts=3
+(d0, <opaque blob A>, 3, -1)
+// Apply for the second txn (the timestamps can be different for each
+// retraction in a txn, but don't need to be)
+(d0, <opaque blob B>, 5, -1)
+(d0, <opaque blob C>, 6, -1)
+(d1, <opaque blob D>, 6, -1)
+```
```

Comment:
Can these retractions cause other transactions to abort due to conflicting timestamps?

Reply:
Yeah, you've correctly anticipated an issue I found the hard way in the WIP integration branch. Even more than aborts, it seems everything gets real panic-y because of the single-envd assumptions.

I think the fix actually works out pretty neatly. We split what is currently "apply" into two steps: "apply" (which only does the compare_and_append_batch and which is the bit to unblock reads) and "tidy" (which retracts the completed apply work). The tidy step then can take a timestamp, which lets us involve the timestamp oracle (though I think in practice we'll want to just sneak it in with the next txn commit to save both the timestamp and the crdb write).

Updated the doc to reflect the above. The "tidy" jargon is a strawman, taking counter-proposals if you have them.

Reply:
There's are all mostly synonyms, but in case you like one better:

- cleanup
- prune
- trim

PostgreSQL refers to a similar process as cleanup (https://www.postgresql.org/docs/current/runtime-config-wal.html#GUC-ARCHIVE-CLEANUP-COMMAND).